### PR TITLE
feat: Add forgejo_personal_access_token resource.

### PR DESCRIPTION
**Pull request by kad-hollac1**
Refs. https://github.com/svalabs/terraform-provider-forgejo/issues/96.

### CODE REVIEW

**Comment by acch** [comment]
Hi @kad-hollac1, thanks much for this valuable contribution!

Looks like acceptance tests are failing for the new data source and resource...

Also, I was wondering if we need to maintain our own custom `stringSliceToList()` / `stringSliceToSet()` functions. Would it be possible to use `types.ListValueFrom()` [0] / `types.SetValueFrom()` [1] framework functions instead, like we already do in other places? [2] This would allow us to reduce the amount of custom code we have to maintain ourselves, in favor of code already provided by the plugin framework...

[0] https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v1.19.0/types#ListValueFrom
[1] https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v1.19.0/types#SetValueFrom
[2] https://github.com/svalabs/terraform-provider-forgejo/blob/4a523f4c25a479a4800a2e3372cca46b738e048e/internal/provider/repository_webhook_resource.go#L69

**Comment by kad-hollac1** [comment]
I'm all in favour of dropping code. ;) I found the branch protection resource as an example and used that, but apparently there was another way. :) I've updated the pull request. The unit tests pass locally, but I can't run the ACC tests on my work laptop. So there may be failures there.

**Comment by kad-hollac1** [comment]
(Do you want me to ask for workflow approval after each new push?)

**Comment by kad-hollac1** [comment]
I've had some troubles with the basic-auth requirement for tokens, but I think I've solved the test setup neatly now. Let me know what you think. :)

**Comment by acch** [comment]
Acc tests still seem to be failing...

**Comment by kad-hollac1** [comment]
Yeah, I totally forgot about the ,,not allowed to specify both a token and a username/password''.

**Comment by acch** [comment]
Hi @kad-hollac1, thank you very much for this valuable contribution! Unfortunately, I'm struggling with understanding the changes to the (test-)setup associated with this PR. Would appreciate if you could document the rationale behind these changes.

With my existing test-setup, ACC tests fail with the following error:

```console
=== RUN   TestAccPersonalAccessTokenResource
    personal_access_token_resource_test.go:15: Step 2/5 error: Error running pre-apply plan: exit status 1

        Error: No value for required variable

          on terraform_plugin_test.tf line 13:
          13: variable "FORGEJO_BASIC_AUTH_USERNAME" {

        The root module input variable "FORGEJO_BASIC_AUTH_USERNAME" is not set, and
        has no default value. Use a -var or -var-file command line argument to
        provide a value for this variable.

        Error: No value for required variable

          on terraform_plugin_test.tf line 16:
          16: variable "FORGEJO_BASIC_AUTH_PASSWORD" {

        The root module input variable "FORGEJO_BASIC_AUTH_PASSWORD" is not set, and
        has no default value. Use a -var or -var-file command line argument to
        provide a value for this variable.
--- FAIL: TestAccPersonalAccessTokenResource (0.46s)
```

Also, I see that you've added a shell-script `internal/provider/env.sh`... what's the purpose of this script?

If I'm not mistaken then you are including the default provider config (`providerConfig`) in your test cases, and also defining an alternate provider config alias ('basicAuth'). Do we really need two provider configs for the same test case? Isn't it possible to alternate the provider config using environment variables?

If this test case really requires username & password, as opposed to API token: have you considered switching to username & password for all test cases? This would eliminate the need to deal with different provider configs...

**Comment by kad-hollac1** [comment]
Yeah, I've jumped some hoops to allow both providers. The env.sh is indeed not needed anymore, I've deleted it.

> If this test case really requires username & password, as opposed to API token: have you considered switching to username & password for all test cases? This would eliminate the need to deal with different provider configs...

I indeed considered switching, but chose this for a few reasons.
- Larger coverage, we can test both the token and the basic-auth routes.
- I assumed that we want to keep token auth as much as possible.

> Isn't it possible to alternate the provider config using environment variables?
Yes, but not per test-case, as far as I know.

**Comment by acch** [comment]
OK, got it. Thanks for the explanation!

Since all test cases seem to be compatible with basic-auth, I'd say we simply switch to that auth method globally - in order to keep things as simple as possible. In the future, if we really do need separate provider blocks for separate test cases, I would suggest to duplicate the const definition [0], and then selectively include one of the two `providerConfig`'s in each test case. This helps to keep the provider config consistent across all resources and test cases.

I already made the required changes locally, if you 'allow edits by maintainers' I can directly push them to your branch...

I'll thoroughly review your actual data source & resource implementation as soon as I can get to it.

[0] https://github.com/svalabs/terraform-provider-forgejo/blob/5c0f3dff18071cdfc30c628c66677336b4ecc62d/internal/provider/provider_test.go#L24

**Comment by kad-hollac1** [comment]
I'd glady allow you to push to the branch, but: https://github.com/orgs/community/discussions/5634. You can upload the patch file?

**Comment by acch** [comment]
> I'd glady allow you to push to the branch, but: https://github.com/orgs/community/discussions/5634. You can upload the patch file?

I wasn't aware of this limitation. Learned something new today, thanks ;-)

Here's the patch (apply with `git am *.patch`):
[0001-chore-use-basic-auth-authentication-method-for-accep.patch](https://github.com/user-attachments/files/30735100/0001-chore-use-basic-auth-authentication-method-for-accep.patch)

File: internal/provider/personal_access_token_data_source.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+// Schema defines the schema for the data source.
+func (d *personalAccessTokenDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "Forgejo personal access token data source.",
+
+		Attributes: map[string]schema.Attribute{
+			"user_id": schema.Int64Attribute{
+				Description: "ID of the user.",
+				Required:    true,
+			},
```
Comment (acch) [inline internal/provider/personal_access_token_data_source.go]: Other, similar data sources - such as GPG keys and SSH keys - reference the owning user by name:
```go
"user": schema.StringAttribute{
	Description: "Name of the user. Changing this forces a new resource to be created.",
```

This is an exception to the "reference other resources by ID" rule, as users are likely to be created outside of Terraform workflows.

Hence, in order to keep things consistent across resources, I'd suggest to reference the owning user by name rather than by ID. This also makes the implementation simpler, because you can remove the call to `getUserByID()`...

File: internal/provider/personal_access_token_resource.go [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+**Note**: Due to an upstream limitation, one cannot create access tokens when authorised with access tokens. Use basic-auth instead.`,
+
+		Attributes: map[string]schema.Attribute{
+			"user_id": schema.Int64Attribute{
+				Description: "ID of the user.",
+				Required:    true,
+				PlanModifiers: []planmodifier.Int64{
+					int64planmodifier.RequiresReplace(),
+				},
+			},
```
Comment (acch) [inline internal/provider/personal_access_token_resource.go]: Other, similar resources - such as GPG keys and SSH keys - reference the owning user by name:

```go
"user": schema.StringAttribute{
	Description: "Name of the user. Changing this forces a new resource to be created.",
```

This is an exception to the "reference other resources by ID" rule, as users are likely to be created outside of Terraform workflows.

Hence, in order to keep things consistent across resources, I'd suggest to reference the owning user by name rather than by ID. This also makes the implementation simpler, because you can remove the call to `getUserByID()`...

File: internal/provider/personal_access_token_resource.go [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+
+	// Use Forgejo client to get personal access token
+	token, diags := getPersonalAccessToken(ctx, r.client, data.UserID.ValueInt64(), data.Name.ValueString())
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Map response body to model
+	data.from(ctx, token)
```
Comment (acch) [inline internal/provider/personal_access_token_resource.go]: You are discarding diagnostics here... should be:

```go
diags = data.from(ctx, token)
resp.Diagnostics.Append(diags...)
if resp.Diagnostics.HasError() {
	return
}
```

File: internal/provider/personal_access_token_resource.go [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+
+// from is a helper function to load an API struct into Terraform data model.
+func (m *personalAccessTokenResourceModel) from(ctx context.Context, t *forgejo.AccessToken) (diags diag.Diagnostics) {
+	if t == nil {
+		return diags
+	}
+
+	m.ID = types.Int64Value(t.ID)
+	m.Name = types.StringValue(t.Name)
+	m.Token = types.StringValue(t.Token)
```
Comment (acch) [inline internal/provider/personal_access_token_resource.go]: If I'm not mistaken then the API only returns the actual token during creation. Any subsequent call will just return "" or nil or something like that.

If we unconditionally overwrite the token in `from()`, that means that any resource refresh (calling `Read()`, calling `from()`) will effectively wipe the saved value...

Instead, I would suggest to move this line out of `from()` and explicitly call it once in `Create()`.

File: internal/provider/personal_access_token_resource.go [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+			}
+		}
+		resp.Diagnostics.AddError("Unable to create personal access token", msg)
+
+		return
+	}
+
+	// Map response body to model
+	diags = data.from(ctx, token)
+	resp.Diagnostics.Append(diags...)
```
Comment (acch) [inline internal/provider/personal_access_token_resource.go:278]: Missing the `HasError()` check:

```go
if resp.Diagnostics.HasError() {
	return
}
```

File: internal/provider/personal_access_token_resource.go [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+		Attributes: map[string]schema.Attribute{
+			"user_id": schema.Int64Attribute{
+				Description: "ID of the user.",
+				Required:    true,
+				PlanModifiers: []planmodifier.Int64{
+					int64planmodifier.RequiresReplace(),
+				},
+			},
+			"name": schema.StringAttribute{
+				Description: "Name of the personal access token.",
```
Comment (acch) [inline internal/provider/personal_access_token_resource.go]: "... Changing this forces a new resource to be created."

File: internal/provider/personal_access_token_resource.go [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+				Required:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"scopes": schema.SetAttribute{
+				Description: "Scopes of the personal access token.",
```
Comment (acch) [inline internal/provider/personal_access_token_resource.go]: "... Changing this forces a new resource to be created."

File: internal/provider/personal_access_token_data_source.go [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+// Metadata returns the data source type name.
+func (d *personalAccessTokenDataSource) Metadata(_ context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_personal_access_token"
+}
+
+// Schema defines the schema for the data source.
+func (d *personalAccessTokenDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "Forgejo personal access token data source.",
```
Comment (acch) [inline internal/provider/personal_access_token_data_source.go:46]: Does the basic-auth API limitation apply to READ as well? If so, a note similar to the resource note should be added to the data source:

> "**Note**: Due to an upstream limitation, one cannot read access tokens when authorized with access tokens. Use basic-auth instead."

File: examples/resources/forgejo_personal_access_token/resource.tf [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  }
+}
+
+variable "test_password" { sensitive = true }
+
+provider "forgejo" {
+  host = "http://localhost:3000"
+
+  username = "admin"
+  password = var.forgejo_password
```
Comment (acch) [inline examples/resources/forgejo_personal_access_token/resource.tf]: This is non-functional, as `var.forgejo_password` hasn't been declared.

I would suggest to keep the provider configuration minimal (no `username`/`password`) and instead repeat the note from the provider docs:

```terraform
provider "forgejo" {
  host = "http://localhost:3000"
  // Due to an upstream limitation, one cannot create access tokens when authorized with access tokens.
  // Use basic-auth instead (FORGEJO_USERNAME / FORGEJO_PASSWORD environment variables).
}
```

File: internal/provider/personal_access_token_resource.go [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+
+	var msg string
+	if res == nil {
+		msg = fmt.Sprintf("Unknown error with nil response: %s", err)
+	} else {
+		tflog.Error(ctx, "Error", map[string]any{
+			"status": res.Status,
+		})
+
+		switch res.StatusCode {
```
Comment (acch) [inline internal/provider/personal_access_token_resource.go:368]: I think the DELETE endpoint also requires basic-auth, so please check for 401 error code.

File: internal/provider/personal_access_token_data_source.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+	if err != nil {
+		var msg string
+		if res == nil {
+			msg = fmt.Sprintf("Unknown error with nil response: %s", err)
+		} else {
+			tflog.Error(ctx, "Error", map[string]any{
+				"status": res.Status,
+			})
+
+			switch res.StatusCode {
```
Comment (acch) [inline internal/provider/personal_access_token_data_source.go:159]: I think the READ endpoint also requires basic-auth, so please check for 401 error code.

**Review by acch — CHANGES_REQUESTED** [review]
Again, thank you VERY much for your contribution @kad-hollac1!

There are a few details that need to be changed, but overall the code is very clean. Great job!

File: internal/provider/personal_access_token_data_source.go (hunk 8 above)
Comment (kad-hollac1) [reply internal/provider/personal_access_token_data_source.go:46]: Nope, it doesn't. :)

File: internal/provider/personal_access_token_resource.go (hunk 10 above)
Comment (kad-hollac1) [reply internal/provider/personal_access_token_resource.go:368]: Nope, it doesn't. :)

File: internal/provider/personal_access_token_data_source.go (hunk 11 above)
Comment (kad-hollac1) [reply internal/provider/personal_access_token_data_source.go:159]: Nope, it doesn't. :)

**Comment by kad-hollac1** [comment]
Regarding DELETE and READ also requiring basic-auth: that is one advantage of the current state, it is very explicit in the tests what needs basic auth and what not. Only creation is limited.

The rationale behind this is that one shouldn't be able to do priviledge escalation. So a token with only user-write rights should not be able to make a token with admin rights. They chose to close the creation endpoint to all tokens instead of checking escalation explicitly.

**Comment by kad-hollac1** [comment]
Thank you for the review! I've addressed all points except the basic-auth/token setup. I do have simplified it, based on your suggestion.

**Comment by kad-hollac1** [comment]
(Please do let me know if you still want to switch to basic auth for everything.)

**Comment by acch** [comment]
Thank you very much for this @kad-hollac1! Yes, you are right: defining both provider configs (by selectively "unsetting" the attributes that are not required for the config) is absolutely the right way. I agree that we should default to token-based auth, unless explicitly required.

The code looks excellent - again, thank you very much for this! I'll merge the changes now, and fix acc tests on the main branch...

Dankuwel!

**Comment by kad-hollac1** [comment]
Thanks again for the thorough review, it is a pleasure contributing to this project! :D

**Comment by acch** [comment]
The pleasure is all mine - keep up the great work!
